### PR TITLE
feat: caching for custom task.

### PR DESCRIPTION
**Pull request by ScrapCodes**
**Which issue is resolved by this Pull Request:** 
Resolves #

**Description of your changes:**

**Environment tested:**

* Python Version (use `python --version`):
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: tekton-catalog/pipeline-loops/go.mod [hunk 1]
```diff
@@ -3,9 +3,13 @@ module github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops
 go 1.13
 
 require (
+	github.com/cenkalti/backoff v2.2.1+incompatible
+	github.com/go-sql-driver/mysql v1.6.0
 	github.com/google/go-cmp v0.5.6
 	github.com/hashicorp/go-multierror v1.1.1
+	github.com/jinzhu/gorm v1.9.16
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/go.mod:10]: Do we have to use `jinzhu/gorm`? or can we use `gorm.io/gorm` instead?

File: tekton-catalog/pipeline-loops/pkg/cache/task_cache_store.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+func (t *TaskCacheStore) Put(entry *model.TaskCache) (*model.TaskCache, error) {
+	if !t.Enabled {
+		return nil, nil
+	}
+	ok := t.db.NewRecord(entry)
+	if !ok {
+		return nil, fmt.Errorf("failed to create a new cache entry, %#v, Error: %v", entry, t.db.Error)
+	}
+	rowInsert := &model.TaskCache{}
+	d := t.db.Create(entry).Scan(rowInsert)
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/pkg/cache/task_cache_store.go:69]: Sorry where do you define the database table here? I'm new to the gorm package so I'm not sure where you are referring to the `task_caches` table name. I only saw `task_caches` during the Get call.

**Comment by Tomcli** [comment]
Thanks @ScrapCodes, I ran this code in my new cluster and here is what I found
1. The new caching package can create the cachedb with `task_caches` table. However, `task_caches` never get updated with new entries even though I ran the pipeline multiple times. Also, I don't see we are referring the `task_caches` table name anywhere outside of Get.
2. We need to check for the label for `pipelines.kubeflow.org/cache_enabled: "true"` in the `Run` CR labels. Otherwise we skip the caching.
3. We want to separate the caching package into its own repo just like the logger package.

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 3]
```diff
@@ -112,16 +118,23 @@ func init() {
 	if err != nil {
 		log.Fatalf("failed to marshal patch bytes in order to cancel: %v", err)
 	}
+	params = db.DBConnectionParameters{}
+	params.LoadDefaults()
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:122]: You also need to run this for the CLI. Right now the tests are falling.

File: tekton-catalog/pipeline-loops/go.mod (hunk 1 above)
Comment (Tomcli) [reply tekton-catalog/pipeline-loops/go.mod:10]: nvm, go-gorm only works for v2 and above. KFP also has this dependency so should be fine.

File: tekton-catalog/pipeline-loops/pkg/cache/task_cache_store_test.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	_ "github.com/mattn/go-sqlite3"
+)
+
+func NewTestingDb() (*gorm.DB, error) {
+	db, err := gorm.Open("sqlite3", ":memory:")
+	if err != nil {
+		return nil, fmt.Errorf("Could not create the GORM database: %v", err)
+	}
+	// Create tables
+	db.AutoMigrate(&model.TaskCache{})
```
Comment (ScrapCodes) [inline tekton-catalog/pipeline-loops/pkg/cache/task_cache_store_test.go:32]: Table is created here.

File: tekton-catalog/pipeline-loops/pkg/cache/db/db_conn_manager.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	// db is safe for concurrent use by multiple goroutines
+	// and maintains its own pool of idle connections.
+	db, err := gorm.Open(driverName, arg)
+	if err != nil {
+		return nil, err
+	}
+	// Create table
+	response := db.AutoMigrate(&model.TaskCache{})
```
Comment (ScrapCodes) [inline tekton-catalog/pipeline-loops/pkg/cache/db/db_conn_manager.go:80]: Table is created here, at this step gorm will generate the necessary sql to create a table with plural table name. e.g. `task_caches`.

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 6]
```diff
@@ -122,7 +122,7 @@ func init() {
 }
 
 func isCachingEnabled(run *v1alpha1.Run) bool {
-	return run.ObjectMeta.Annotations["pipelines.kubeflow.org/cache_enabled"] == "true"
+	return run.Labels["pipelines.kubeflow.org/cache_enabled"] == "true"
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:125]: ObjectMeta.Labels?

**Comment by ScrapCodes** [comment]
On resubmitting the same pipeline run.
```
$ k get runs
NAME                        SUCCEEDED   REASON     STARTTIME   COMPLETIONTIME
pr-loop-example-loop-task   True        CacheHit   28s         28s
```

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 7]
```diff
@@ -112,16 +117,28 @@ func init() {
...
+func isCachingEnabled(run *v1alpha1.Run) bool {
+	return run.Labels["pipelines.kubeflow.org/cache_enabled"] == "true"
 }
 
 // ReconcileKind compares the actual state with the desired, and attempts to converge the two.
 // It then updates the Status block of the Run resource with the current status of the resource.
 func (c *Reconciler) ReconcileKind(ctx context.Context, run *v1alpha1.Run) pkgreconciler.Event {
 	var merr error
 	logger := logging.FromContext(ctx)
+	err := CacheStore.Connect(params)
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:133]: Can we also have an environment variable to disable the cache store for pipelineloop? I think the product team needs some time to migrate this behavior, so is better to have some flag that can turn it on/off.

**Comment by Tomcli** [comment]
The SDK didn't add the default caching labels for the pipeline loop run spec. Need to open an issue to handle this

**Comment by ScrapCodes** [comment]
@Tomcli Thank you for all the help ! Please take a look again.

File: tekton-catalog/cache/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
+# Cache: Reuse the results from previous execution for custom tasks.
+
+### How To
+
+1. Setup.
+```go
+
+import (
+    "fmt"
+	"time"
```
Comment (Tomcli) [inline tekton-catalog/cache/README.md:10]: Fix typo
```suggestion
    "time"
```

**Comment by Tomcli** [comment]
Thanks @ScrapCodes, just a small typo

**Comment by Tomcli** [comment]
thanks @ScrapCodes, this is good for the initial version and we will get more feedbacks from the internal to further improve this.

**Pull request merged**